src/protagonist/Orchestrator/Infrastructure/IIIF/IIIFManifestBuilder.cs

```diff
@@ -0,0 +1,158 @@
+            {
+                new(Language, "Title", "Created by DLCS"),
+                new(Language, "Generated On", DateTime.UtcNow.ToString("u"))
+            }
+        };
+        
+        manifest.EnsurePresentation3Context();
+        if (anyAssetRequireAuth)
+        {
+            logger.LogTrace(
+                "ManifestId {ManifestId} has at least 1 asset requiring auth - adding Auth2 context + services",
+                manifestId);
+            manifest.EnsureContext(IIIFAuth2.Constants.IIIFAuth2Context);
+
+            // Add the AuthAccessServices to the manifest services collection.
+            // Individual ImageServices will contain ProbeService and reference this accessService
+            var accessServices = GetDistinctAccessServices(probeServices);
+            manifest.Services = accessServices;
+        }
+        
+        var canvases = await canvasFactory.CreateV3Canvases(assets, customerPathElement, probeServices);
+        manifest.Items = canvases;
+        manifest.Thumbnail = canvases.FirstOrDefault(c => !c.Thumbnail.IsNullOrEmpty())?.Thumbnail;
+        
+        return manifest;
+    }
+
+    public async Task<IIIF2.Manifest> GenerateV2Manifest(List<Asset> assets, CustomerPathElement customerPathElement,
+        string manifestId, string label, string sequenceRoot, CancellationToken cancellationToken)
+    {
+        var manifest = new IIIF2.Manifest
+        {
+            Id = manifestId,
+            Label = new MetaDataValue(label),
+            Metadata = new List<IIIF2.Metadata>
+            {
+                new()
+                {
+                    Label = new MetaDataValue("Title"),
+                    Value = new MetaDataValue("Created by DLCS")
+                } ,
+                new()
+                {
+                    Label = new MetaDataValue("Generated On"),
+                    Value = new MetaDataValue(DateTime.UtcNow.ToString("u"))
+                }   
+            }
+        };
+        
+        manifest.EnsurePresentation2Context();
+        
+        var canvases = await canvasFactory.CreateV2Canvases(assets, customerPathElement);
+        var sequence = new IIIF2.Sequence
+        {
+            Id = string.Concat(sequenceRoot, "/sequence/0"),
+            Label = new MetaDataValue("Sequence 0"),
+        };
+        sequence.Canvases = canvases;
+        manifest.Thumbnail = canvases.FirstOrDefault(c => !c.Thumbnail.IsNullOrEmpty())?.Thumbnail;
+        manifest.Sequences = sequence.AsList();
+
+        return manifest;
+    }
+
+    private async Task<Dictionary<AssetId, AuthProbeService2>?> GetProbeServices(IReadOnlyCollection<Asset> assets, CancellationToken cancellationToken)
+    {
+        var assetsRequiringAuth = assets.Where(a => a.RequiresAuth && !string.IsNullOrEmpty(a.Roles)).ToList();
+
+        var assetsRequiringAuthCount = assetsRequiringAuth.Count;
+        if (assetsRequiringAuthCount == 0) return null;
```

Comment:
what's the reason for returning null over an empty dictionary?

Reply:
The vast majority of assets won't have any items requiring auth (ie `assetsRequiringAuthCount == 0`) so it avoids instantiating a new object in these cases, only for it to be disposed of.